Feature/sync team member to firebase

Rup-Narayan-Rajbanshi opened the pull request:
## Depends on
- https://github.com/mapswipe/mapswipe-firebase/pull/19

## Changes
- add team members when create or update team.
- add base class for firebase sync feature
- modify organization sync to firebase according to new class.
- add management command to create dummy contributor user.
- add link to view contributor users in team row.
- allow to update only team in contributor admin form.

## This PR doesn't introduce any:

- [x] temporary files, auto-generated files or secret keys
- [x] n+1 queries
- [x] flake8 issues
- [x] `print`
- [x] typos
- [x] unwanted comments

## This PR contains valid:

- [ ] tests
- [ ] permission checks (tests here too)
- [ ] translations

Rup-Narayan-Rajbanshi committed:
chore(firebase): add common func for push_to_firebase

  - Add field old id in teams
  - Refactor project to accept base firebase resources.

Rup-Narayan-Rajbanshi committed:
chore(contributor-team): sync team member to firebase when team update or create

tnagorra commented on main/config.py:
```diff
@@ -50,6 +50,10 @@ def organization(organization_id: str | int):
             return f"/v2/teams/{contributor_team_id}/"
 
+        @staticmethod
+        def users(user_id: str | int):
```
```suggestion
        def contributor_user(user_id: str | int):
```

tnagorra commented on apps/project/firebase.py:
```diff
@@ -1,97 +1,58 @@
+    def handle_new_object_on_firebase(self, model_obj: Organization, fb_reference: FbReference):
+        organization_data = firebase_models.FbOrganisation(
+            name=model_obj.name,
+            nameKey="",
```
Let's add a FIXME or NOTE why nameKey is empty

tnagorra commented on apps/project/firebase.py:
```diff
@@ -1,97 +1,58 @@
+            value={
+                **firebase_utils.serialize(organization_data),
+            },
+        )
```
We do no need to destructure the response of serialize

tnagorra commented on apps/project/firebase.py:
```diff
@@ -1,97 +1,58 @@
+            value=firebase_utils.serialize(
+                firebase_models.FbOrganisation(
+                    name=model_obj.name,
+                    nameKey="",
```
Let's add a FIXME or NOTE why nameKey is empty

tnagorra commented on apps/project/firebase.py:
```diff
@@ -1,97 +1,58 @@
 
 
-class InvalidOrganizationPushException(Exception): ...
+class FirebaseOrganization(FirebasePush[Organization]):
```
```suggestion
class FirebaseOrganizationPush(FirebasePush[Organization]):
```

tnagorra commented on apps/contributor/admin.py:
```diff
@@ -57,7 +57,14 @@ def save_model(self, request, obj, form, change):
-        transaction.on_commit(lambda: push_contributor_team_to_firebase.delay(obj.pk))
+        transaction.on_commit(lambda: FirebaseContributorTeam.task.delay(obj.id))
+
+    @typing.override
+    def save_related(self, request, form, formsets, change):
+        super().save_related(request, form, formsets, change)
+        contributor_users = form.instance.user.all()
+        for user in contributor_users:
+            transaction.on_commit(lambda user_id=user.pk: firebase_contributor_user.delay(user_id))
```
We do not need to save related models.

tnagorra commented on apps/contributor/firebase.py:
```diff
@@ -1,50 +1,104 @@
+    model_obj: ContributorUser
+    model = ContributorUser
+
+    @typing.override
+    def handle_new_object_on_firebase(self, model_obj: ContributorUser, fb_reference: FbReference):
+        team_id = str(model_obj.team.id) if model_obj.team else ""
+        user_data = firebase_models.FbUser(
+            userName=model_obj.username,
+            username="",
+            userNameKey="",
+            usernameKey="",
+            teamId=team_id,
+            created=timezone.now(),
+        )
+        fb_reference.set(
+            value=firebase_utils.serialize(user_data),
+        )
```
We should not create contributor user on firebase.

tnagorra commented on apps/contributor/firebase.py:
```diff
@@ -1,50 +1,104 @@
+            value=firebase_utils.serialize(user_data),
+        )
+
+    @typing.override
+    def handle_object_update_on_firebase(self, model_obj: ContributorUser, fb_reference: FbReference):
+        team_id = str(model_obj.team.id) if model_obj.team else ""
+        fb_reference.update(
+            value=firebase_utils.serialize(
+                firebase_models.FbUser(
+                    userName=model_obj.username,
+                    username="",
+                    userNameKey="",
+                    usernameKey="",
+                    teamId=team_id,
+                    created=timezone.now(),
+                ),
+            ),
+        )
```
We should only update "teamId"

tnagorra commented on apps/contributor/firebase.py:
```diff
@@ -1,50 +1,104 @@
+        FirebaseContributorTeam(obj_id).push()
+
+
+class FirebaseContributorUser(FirebasePush[ContributorUser]):
```
ContributorUser cannot be a FirebaseResource unless we sort out the old_id and canonical_id to also include the user's firebase id.

tnagorra committed:
feat(admin): remove inline contributor user create on team

- disable adding/deleting contributor user
- add a reusable mixing for readonly admin interface

tnagorra committed:
fix(firebase): update firebase push logic for contributor user

- only allow editing teamId of contributor user during firebase update
- disallow creating contributor user on backend
- fix issue with firebase key for tasks

Rup-Narayan-Rajbanshi committed:
chore(admin): add management command to create dummy contributor user

- Add link to view contributor users in team row.
- Allow to update only team in contributor admin form.

tnagorra committed:
feat(firebase): introduce firebase_id to replace canonical_id

- remove canonical_id and use firebase_id instead
- rename legacy_task_id to firebase_id
- rename legacy_group_id to firebase_id
- add resuable models: FirebasePushModel, FirebasePullModel

tnagorra commented on apps/contributor/management/commands/create_contributor_users.py:
```diff
@@ -0,0 +1,38 @@
+        )
+        team_a, team_b = ContributorTeamFactory.create_batch(2, **user_resources)
+        for team in [team_a, team_b]:
+            transaction.on_commit(lambda team_id=team.pk: FirebaseContributorTeam.task.delay(team_id))
```
Let's do it directly.

tnagorra commented on apps/contributor/management/commands/create_contributor_users.py:
```diff
@@ -0,0 +1,38 @@
+
+        ContributorUserFactory.create_batch(5, team=team_a)
+        ContributorUserFactory.create_batch(5, team=team_b)
+        ContributorUserFactory.create_batch(20)
```
We also need to push these to firebase.

tnagorra commented on apps/contributor/factories.py:
```diff
@@ -20,7 +21,7 @@ class ContributorUserFactory(DjangoModelFactory):
         model = ContributorUser
 
-    user_id = factory.Sequence(lambda n: f"unique-contributor-user-id-{n}")
+    user_id = factory.LazyAttribute(lambda _: f"contributor-user-id-{uuid.uuid4().int % 1000000}")
```
Let's just directly use ULID.

tnagorra commented on apps/contributor/admin.py:
```diff
@@ -59,6 +71,10 @@ def save_model(self, request, obj, form, change):
         super().save_model(request, obj, form, change)  # type: ignore[reportAttributeAccessIssue]
         transaction.on_commit(lambda: FirebaseContributorTeam.task.delay(obj.id))
 
+    def view_contributor_users(self, obj):
+        url = reverse("admin:contributor_contributoruser_changelist") + f"?team__id__exact={obj.id}"
+        return format_html('<a href="{}">View Contributors</a>', url)
```
```suggestion
    def view_team_members(self, obj):
        url = reverse("admin:contributor_contributoruser_changelist") + f"?team__id__exact={obj.id}"
        return format_html('<a href="{}">View Team Members</a>', url)
```

Rup-Narayan-Rajbanshi committed:
chore(firebase): sync team members to firebase while removing team member

tnagorra committed:
chore(migrations): add a merge migration
